### DISCUSSION TITLE
I'm confused

### DISCUSSION THREAD

**Discussion by TheRealJosephTheMan** [mrdbourke/pytorch-deep-learning#1015] [Q&A]
so in the video section 54 timestamp 8:02:38 he puts `model_1.to(device)
next(model_0.parameters())` it didnt worked it makes sense why it didnt work because i didnt create model 1 but i didnt see him create model 1 what do i do? heres colab link: https://colab.research.google.com/drive/1gKl1jetrdNoMU5Brwbz6DFh20BfW6udT?usp=sharing

**Comment by ghost** [comment]
Based on your google colab code , it seems that you're switching from a custom linear regression model (`model_0`) to a more straightforward linear model (`model_1`) using PyTorch's `nn.Linear` layer. You encountered an issue with `model_1` not being created correctly, and there's a reference to using `model_1.to(device)` which suggests moving the model to a GPU if available.

### Addressing the Specific Issue:

1. **Initialization of `model_1`:**
   Ensure that `model_1` is properly initialized before using it. Your code defines `model_1` as an instance of `LinearRegressionModelV2`:

   ```python
   class LinearRegressionModelV2(nn.Module):
     def __init__(self):
       super().__init__()
       self.linear_layer = nn.Linear(in_features=1, out_features=1)

     def forward(self, x: torch.Tensor) -> torch.Tensor:
         return self.linear_layer(x)
   ```

   This model should be created as follows:
   ```python
   model_1 = LinearRegressionModelV2()
   ```

2. **Moving Model to Device:**
   Before you use `model_1.to(device)`, ensure that `device` is correctly set:
   ```python
   device = "cuda" if torch.cuda.is_available() else "cpu"
   ```
   Move the model to the device:
   ```python
   model_1.to(device)
   ```

3. **Data on the Same Device:**
   Ensure that your data (`X_train`, `y_train`, `X_test`, `y_test`) is also moved to the same device:
   ```python
   X_train, y_train = X_train.to(device), y_train.to(device)
   X_test, y_test = X_test.to(device), y_test.to(device)
   ```

4. **Verify Model Parameters:**
   After moving the model to the device, verify the device of the model parameters:
   ```python
   next(model_1.parameters()).device
   ```

5. **Training Loop:**
   Make sure your training loop runs correctly and uses the model on the correct device. Here’s a cleaned-up version of the training loop:
   ```python
   torch.manual_seed(42)
   epochs = 200
   for epoch in range(epochs):
     model_1.train()
     y_pred = model_1(X_train)
     loss = torch.l1_loss(y_pred, y_train)
     optimiser.zero_grad()
     loss.backward()
     optimiser.step()
     model_1.eval()
     with torch.inference_mode():
       test_pred = model_1(X_test)
       test_loss = torch.l1_loss(test_pred, y_test)
       if epoch % 10 == 0:
         print(f'Epoch: {epoch} | Loss: {loss.item()} | Test loss: {test_loss.item()}')
   ```

6. **Debugging Output:**
   Print the `model_1` state_dict and other outputs to ensure everything is working as expected:
   ```python
   print(model_1.state_dict())
   ```

7. **Plotting Predictions:**
   Ensure that when plotting predictions, you detach tensors from the computation graph and convert them to NumPy arrays:
   ```python
   with torch.inference_mode():
     y_preds_new = model_1(X_test)
   y_preds_new_np = y_preds_new.detach().cpu().numpy()
   plot_predictions(predictions=y_preds_new_np)
   ```


I hope that could help :)

**Comment by TheRealJosephTheMan** [comment]
thanks
